core/tests/test_utils.py

```diff
@@ -2454,25 +2457,35 @@ def assertRaisesRegexp(  # pylint: disable=keyword-arg-before-vararg
 class AppEngineTestBase(TestBase):
     """Base class for tests requiring App Engine services."""
 
-    # We can't instantiate the stub in setUp because the overrided
-    # method run() executes before setUp() and run() requires the memory
-    # cache stub.
-    def __init__(self, *args, **kwargs):
-        super(AppEngineTestBase, self).__init__(*args, **kwargs)
+    def setUp(self):
         self.memory_cache_services_stub = MemoryCacheServicesStub()
         self.taskqueue_services_stub = TaskqueueServicesStub(self)
 
-    def _delete_all_models(self):
-        """Deletes all models from the NDB datastore."""
-        datastore_services.delete_multi(
-            datastore_services.query_everything().iter(keys_only=True))
+        with contextlib2.ExitStack() as stack:
+            stack.enter_context(self.swap(
+                platform_taskqueue_services, 'create_http_task',
+                self.taskqueue_services_stub.create_http_task))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'flush_cache',
+                self.memory_cache_services_stub.flush_cache))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_multi',
+                self.memory_cache_services_stub.get_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'set_multi',
+                self.memory_cache_services_stub.set_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_memory_cache_stats',
+                self.memory_cache_services_stub.get_memory_cache_stats))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'delete_multi',
+                self.memory_cache_services_stub.delete_multi))
+
+            self._stack = stack.pop_all()
```

Comment:
I don't really understand what's going on here, why are we popping everything immediately? (And why do we need a stack and a self._stack?)

Maybe try to use naming and comments to explain better what is going on here.

Reply:
At what point are we running code with the swaps in effect? That is not clear to me -- I would have expected it to be "within" the stack somehow, but I can't find where that is happening.

If this is applied "globally", then I think the thing that is mostly confusing me at the moment is all the machinations that involve transferring stuff between the stacks. It feels kind of hacky.

Probably because of the above, I find the existing strategy easier to understand, but maybe I'm just used to it. Let's also see what the other reviewers of the PR say.

Reply:
`stack.enter_context` is where the swaps are put into effect. After that, they are applied "globally" because we have entered its context. This is necessary because we want to exit the context (globally) from the `tearDown` method.

I disagree in finding it "hacky," this is exactly how it's designed to be used.

Also, note that `ExitStack` is actually a part of the standard Python 3 library ([docs](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack)). `contextlib2` is just a backport.

Reply:
Do you have doc examples that show this stack transferring thing is standard? The ones I found were simpler, but I may have missed it. (I'm happy to defer if this is a standard paradigm, but it would help to see evidence.)

My objection is not to the import of contextlib2 but the way it's being used.

Reply:
The [`pop_all`](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack.pop_all) documentation demonstrates how it is used to move the operation of closing contexts to a function. The same pattern is being used here.

Reply:
Thanks for the reference!

What's your take on doing something similar here, i.e. store only self.close_files and execute it in tearDown()?

Reply:
I actually think it's even harder to explain, since it depends on bound method references which is a pretty advanced concept. I think the code as-is does its job in a readable-enough way.

Reply:
OK. I suggest you put the question to the other reviewers, then -- I have a preference the other way, but am happy to be overruled.

Reply:
(I should also point out that there's a principle behind my suggestion, which is to store the minimal context possible. It's similar to how we try to pass ids around rather than large objects.)

Reply:
Would you be more comfortable if I keep the `run` method and use an exit stack there instead?

Reply:
OK, I've decided to bring `run()` back in but use `ExitStack` to manage the swaps. WDYT?